File: LeetCode/0053-Maximum Subarray/main.cpp
//
// Created by 薛智钧 on 2020/3/27.
//
#include <vector>
#include <iostream>
#include <cstddef>
using namespace std;

/*
 * Given an integer array nums,
 * find the contiguous subarray (containing at least one number)
 * which has the largest sum and return its sum.

Example:

Input: [-2,1,-3,4,-1,2,1,-5,4],
Output: 6
Explanation: [4,-1,2,1] has the largest sum = 6.
 * */

/*
 * 定义两个变量 res 和 curSum，其中 res 保存最终要返回的结果，即最大的子数组之和，curSum 初始值为0，
 * 每遍历一个数字 num，比较 curSum + num 和 num 中的较大值存入 curSum，
 * 然后再把 res 和 curSum 中的较大值存入 res，以此类推直到遍历完整个数组，可得到最大子数组的值存在 res 中，代码如下：
 * */
class Solution {
public:
    int maxSubArray(vector<int>& nums) {
        int max_global = INT_MIN, max_local = 0;
        for (int num : nums){
            // 将当前数字和（当前数字＋之前subArray的和）比大小
            // 若当前数字较大，则变更滑动窗口变更为当前数字为起点
            max_local = max(max_local + num, num);
            max_global = max(max_global, max_local);
        }
        return max_global;
    }
};

File: LeetCode/0053-Maximum Subarray/main.py
"""  
Given an integer array nums, find the contiguous subarray (containing at least one number) 
which has the largest sum and return its sum.

Follow up: If you have figured out the O(n) solution, try coding another solution 
using the divide and conquer approach, which is more subtle.


Example 1:

Input: nums = [-2,1,-3,4,-1,2,1,-5,4]
Output: 6
Explanation: [4,-1,2,1] has the largest sum = 6.
"""


"""  
Greedy search

用 res 保存全局连续最大值，cur_sum 保存局部连续最大值
从左向右遍历，比较 cur_sum + num 和 num 的大小，若cur_sum + num < num，则舍弃之前的 cur_sum
使得 cur_sum = num
"""
class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        res, cur_sum = float('-inf'), 0
        
        for num in nums:
            cur_sum = max(cur_sum + num, num)
            res = max(res, cur_sum)
            
        return res

    # 动态规划
    def maxSubArray2(self, nums: List[int]) -> int:
        res = nums[0]
        for i in range(1, len(nums)):
            nums[i] = max(nums[i] + nums[i-1], nums[i])
            res = max(res, nums[i])
        
        return res